### PR TITLE
Add source location to invalid_source_file_exceptiont

### PR DESCRIPTION
**Pull request by tautschnig**
This enables proper use in the C front-end in a way that is compatible
with the compiler-specific message handlers in goto-cc: those message
handlers generate source-location output that mimics that of the
compiler being emulated.

As future work it is thus possible to remove all "throw 0" instances
from the C front-end.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/literals/convert_character_literal.cpp
Comment (chrisr-diffblue) [inline src/ansi-c/literals/convert_character_literal.cpp]: Nice :-)

File: src/ansi-c/literals/convert_character_literal.cpp
Comment (chrisr-diffblue) [inline src/ansi-c/literals/convert_character_literal.cpp:105]: Also a nice catch!

File: src/goto-programs/initialize_goto_model.cpp
Comment (chrisr-diffblue) [inline src/goto-programs/initialize_goto_model.cpp]: Since this is a (slight) change in semantics for this function, such that it now _requires_ the `function` option to be set (previously the function name was passed in as an argument) - is it worth adding a CHECK_RETURN here?

File: src/util/exception_utils.h
Comment (chrisr-diffblue) [inline src/util/exception_utils.h:171]: 👍

**Review by chrisr-diffblue — APPROVED** [review]
That looks great to me, and is what we should have done when we first added the structured exceptions. Some nice code cleanups/refactoring along the way too.

File: src/goto-programs/initialize_goto_model.cpp
Comment (tautschnig) [reply src/goto-programs/initialize_goto_model.cpp]: Good point! I've added the suggested `CHECK_RETURN`.

**Review by chris-ryder — APPROVED** [review]
Previously reviewed and approved with my Diffblue account, re-approving with my non-Diffblue account to keep the CODEOWNER review.

**Review by martin-cs — APPROVED** [review]
Yes.  I don't think this breaks any scripting.

File: src/util/typecheck.cpp
Comment (peterschrammel) [inline src/util/typecheck.cpp:41]: 🎉

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
